pmm-app/src/shared/components/helpers/api.ts

```diff
@@ -10,17 +10,13 @@ export class ApiRequest {
     });
   }
 
-  async get<T, B>(path: string, query?: { params: B; cancelToken?: CancelToken }): Promise<void | T> {
+  async get<T, B>(path: string, query?: { params: B; cancelToken?: CancelToken }): Promise<T> {
```

Comment:
When the promise throws, isn't that the case when it returns `void`?

Reply:
When it throws the execution moves to the catch block so it doesn't actually return anything. If there's no catch block the browser will handle the error but still the code after won't be executed so there's no need to expect void as return.

Made this snippet to try to explain it: https://stackblitz.com/edit/js-rxurye?file=index.js

Reply:
What I see from the code is that `.catch(e)` statement is always there and it always re-throws the error. This means that, although there is no explicit `return` statement, the engine kicks in and returns `void` (or, more precisely, `unknown` in latest versions of typescript).

I asked because leaving only `Promise<T>` would not correspond to what's happening in reality, which is IMO `Promise<T | unknown>`.

My point is: absence of a return statement does not mean there is no `return`.

My other point is: the `.catch` statement seems to be useless if the only thing it does is re-throwing the error.

Reply:
The catch is not just re throwing the error, it is showing the error message.

Can you please show an use case where the code after the get will receive `unknown`?

Reply:
This is basically the definition of `throw`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/throw
_Execution of the current function will stop (the statements after throw won't be executed)_

So the get will either return `T` or it will `throw` it won't ever return `unknown`

Reply:
> The catch is not just re throwing the error, it is showing the error message.
> 
> Can you please show an use case where the code after the get will receive `unknown`?

This is where. 
https://github.com/percona/grafana-dashboards/pull/1297/files#diff-cd8788d8824d855138634130187cbbb636faa9ca5e17e9fb381f43fc8d38b290R19

Reply:
That's where it throws, so it does not return. Also that's inside the API helper function, it's not affected by the return type.

Reply:
If we were putting a wrong return type typescript compiler would warn us.

Reply:
I think you are right. I was pointing to a wrong solution, something that has to do with type inference which is not what we are trying to do )))